File: kotlin/example/src/test/kotlin/com/example/PlaybackConsumerTest.kt
package com.example

import com.paulhammant.servirtium.vcr.Outcome
import com.paulhammant.servirtium.vcr.kotlin.playback
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.nio.file.Paths
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

/**
 * Third-party consumer test: drives the INSTALLED Kotlin DSL jar
 * (servirtium-vcr-kotlin, resolved from ~/.m2) through its `playback { }` DSL,
 * replaying the canonical tape. The engine .so is discovered zero-config from
 * the transitive servirtium-vcr jar's /native/<rid>/ resource — no
 * SERVIRTIUM_VCR_LIB, no source tree.
 */
class PlaybackConsumerTest {

    @Test
    fun `replays canonical tape via the installed kotlin dsl`() {
        val location = Class.forName("com.paulhammant.servirtium.vcr.kotlin.Servirtium")
            .protectionDomain.codeSource.location.toString()
        assertTrue(location.endsWith(".jar"), "expected the installed kotlin jar, got $location")

        val tape = Paths.get(
            javaClass.getResource("/tapes/single_get.md")!!.toURI()
        ).toString()

        playback(tape) { port(0) }.use { vcr ->
            val resp = HttpClient.newHttpClient().send(
                HttpRequest.newBuilder(URI.create(vcr.baseUrl() + "/ok")).GET().build(),
                HttpResponse.BodyHandlers.ofString(),
            )
            assertEquals(200, resp.statusCode())
            assertEquals("ok-body", resp.body())
            assertEquals(Outcome.OK, vcr.lastKind())
        }
    }
}

File: scala/example/src/test/scala/com/example/PlaybackConsumerTest.scala
package com.example

import com.paulhammant.servirtium.vcr.Outcome
import com.paulhammant.servirtium.vcr.scala.Servirtium
import java.net.URI
import java.net.http.{HttpClient, HttpRequest, HttpResponse}
import java.nio.file.Paths
import org.junit.jupiter.api.Assertions.{assertEquals, assertTrue}
import org.junit.jupiter.api.Test

/**
 * Third-party consumer test: drives the INSTALLED Scala helper jar
 * (servirtium-vcr-scala, resolved from ~/.m2) via `Servirtium.playback`,
 * replaying the canonical tape. The engine .so is discovered zero-config from
 * the transitive servirtium-vcr jar resource — no SERVIRTIUM_VCR_LIB.
 */
class PlaybackConsumerTest:

  @Test def replaysCanonicalTapeFromInstalledScalaJar(): Unit =
    val location =
      Servirtium.getClass.getProtectionDomain.getCodeSource.getLocation.toString
    assertTrue(location.endsWith(".jar"), s"expected the installed scala jar, got $location")

    val tape = Paths.get(getClass.getResource("/tapes/single_get.md").toURI).toString

    val vcr = Servirtium.playback(tape)(_.port(0))
    try
      val resp = HttpClient.newHttpClient.send(
        HttpRequest.newBuilder(URI.create(vcr.baseUrl() + "/ok")).GET().build(),
        HttpResponse.BodyHandlers.ofString(),
      )
      assertEquals(200, resp.statusCode())
      assertEquals("ok-body", resp.body())
      assertEquals(Outcome.OK, vcr.lastKind())
    finally vcr.close()
